cdk: Enhance CDK implementation with CLI tool, bug fixes, and improved configuration

Belhoussine opened the pull request:
Issue #, if available:
N/A

Description of changes:
This PR builds upon the previous CDK implementation by adding several enhancements and improvements:

Key additions:
- Implemented new 'acmne-cli' command-line interface tool for easier deployment
- Added advanced usage documentation for both the CLI tool and CDK CLI

Bug fixes and improvements:
- Fixed httpd configuration bug on Amazon Linux 2023
- Resolved props.isPrivate boolean validation issue
- Added default PCA ARN string to prevent breaking the `cdk bootstrap` initialization.
- Fixed user-data script for additional certificate configuration (`/etc/hosts` for private certificates)
- Introduced certificate validation type configuration in Certificate Stack

Code organization and cleanup:
- Renamed 'streamline' folder to 'cdk' for better clarity
- Centralized default resource names to avoid duplication
- Updated default configuration values for serverType and region
- Improved region handling using props.env.region in RoleStack

Documentation updates:
- Added comprehensive README documentation for acmne-cli tool usage
- Included advanced usage guide for CDK CLI
- Updated documentation to reflect new 'cdk' folder structure

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Belhoussine committed:
cdk: Fix httpd bug on AL2023 - resolve missing (undocumented) httpd-acm.conf directives and edit openssl.cnf

Belhoussine committed:
cdk: Update default-config.ts default values for serverType and region

Belhoussine committed:
cdk: Implement 'acmne-cli' CLI tool

Belhoussine committed:
cdk: Store default resource names one level higher to avoid duplication

Belhoussine committed:
cdk: Use props.env.region instead of an explicit region prop in RoleStack

Belhoussine committed:
cdk: Add user-data script for additional configuration if the certificate is private

Belhoussine committed:
cdk: Update README - documentation for acmne-cli tool

Belhoussine committed:
cdk: Add advanced usage guide using CDK CLI

Belhoussine committed:
cdk: Add default PCA ARN string in default-config (prevents breaking initial cdk bootstrap)

Belhoussine committed:
cdk: Fix bug in props.isPrivate boolean - explicitely check existence

Belhoussine committed:
cdk: Add Validation Type prop in Certificate Stack

Belhoussine committed:
cdk: Rename 'streamline' folder to 'cdk'

Belhoussine committed:
cdk:  Update README to point to new 'cdk' folder

meerd commented on cdk/README.md:
````diff
@@ -58,7 +58,7 @@ The app consists of **three** main CDK stacks:
 ### Installation
 ```bash
 git clone <repository-url>
````
The top-level README.md should also briefly mention installing ACM4NE with CDK and include a link to point this README file.

meerd commented on streamline/docs/cdk-usage.md:
```diff
@@ -0,0 +1,160 @@
+# CDK Usage Guide - ACM for Nitro Enclaves Streamline
```
Should we move this file to aws-nitro-enclaves-acm/docs/? The top-level document directory.

meerd commented on streamline/src/bin/cli.ts:
```diff
@@ -0,0 +1,85 @@
+const program = new Command();
+
+program
+  .name('acmne-cli')
```
We should reconsider the CLI's name. This utility is intended for installation. Maybe `setup-tool`? Please see the current tools in this repository: https://github.com/Belhoussine/aws-nitro-enclaves-acm/tree/main/tools

In any case, The tool needs to support both resource deployment and cleanup operations. Such as: 

`acmne-cli setup --setup-name my_test_setup_1 --certificate ...`
`acmne-cli destroy --stack-name my_test_setup_1`

Feel free to propose better naming. :)

Belhoussine replied:
Done!
Slightly changed the naming 
`setup-tool deploy --setup-name myTestSetup --aws-account-id ...`
`setup-tool destroy --setup-name myTestSetup --instance-stack (--role-stack, --certificate-stack or --all)`

meerd commented on streamline/src/bin/cli.ts:
```diff
@@ -0,0 +1,85 @@
+  // Instance config
+  .option('-i, --instance-name <string>', 'Instance name')
+  .option('-k, --key-pair-name <string>', 'Key pair name')
+  .option('-s, --server-type <string>', 'Server type (NGINX or APACHE)')
```
I'd suggest using `--web-server-type` for clarity.

meerd commented on streamline/src/bin/cli.ts:
```diff
@@ -0,0 +1,85 @@
+  .option('-m, --ami-type <string>', 'AMI type (AL2 or AL2023)')
+  // General config
+  .option('-a, --aws-region <string>', 'AWS region')
+  .option('-u, --aws-account <string>', 'AWS account');
```
Maybe `--aws-account-id`?

meerd commented on streamline/src/bin/cli.ts:
```diff
@@ -0,0 +1,85 @@
+  // Role config
+  .option('-r, --role-name <string>', 'Role name')
+  // Instance config
+  .option('-i, --instance-name <string>', 'Instance name')
```
Are there plans to support deployment to existing EC2 instances?

Belhoussine replied:
The current limitation where CDK cannot import existing EC2 instances ([aws-cdk/issues/3174](https://github.com/aws/aws-cdk/issues/3174#issuecomment-586907944)) presents a challenge. 
Rather than exploring complex workarounds (like SDK+SSM integration or CDK library modifications), I would currently opt for a pragmatic hybrid approach:

1. Deploy `certificateStack` and `roleStack` via CDK (using our current provided solution)
2. Attach the created instance profile via AWS CLI:
   ```bash
   aws ec2 associate-iam-instance-profile \
       --region <region> \
       --instance-id <instance_id> \
       --iam-instance-profile Name=AcmneInstanceProfile
    ```
3. Configure web server manually per ACM4NE documentation ([Step 7](https://docs.aws.amazon.com/enclaves/latest/user/install-acm.html#config-web-server))

This solution still deals with most of the cumbersome / error-prone manual steps while working within CDK's constraints. It also avoids introducing additional dependencies or maintenance overhead that would come with more complex alternatives.

Belhoussine committed:
cdk: Add support for ARM CPU type in InstanceStack AMI creation

Belhoussine committed:
cdk: Rework CLI tool - Implement support for deploy & destroy (Accomodate CDK limitations)

cdk: Change the naming for the CLI tool and the CLI options & Add default option values (prevents errors in cdk destroy)

cdk: Add stackName as part of the config - allows for simple integration of setupName and consistency for the destroy subcommand

cdk: Move individual stack validation from stacks into ConfigValidator - prevents CDK destroy errors via CLI

cdk: Rework CLI tool - Implement support for deploy & destroy (Accomodate CDK limitations)

Belhoussine committed:
cdk: Move the instance profile generation from the instanceStack to the roleStack - step towards supporting deployment to existing instances

Belhoussine committed:
cdk: Update README - Add setup-tool documentation

foersleo commented on streamline/src/assets/user-data-scripts/AL2023/apache-conf.sh:
```diff
@@ -3,14 +3,17 @@
 sudo dnf install aws-nitro-enclaves-acm -y
 
-# Generate the /etc/nitro_enclaves/acm.yaml file
+# # Generate the /etc/nitro_enclaves/acm.yaml file
```
nit: Why the double `#` Similarly below

foersleo commented on streamline/src/assets/user-data-scripts/AL2023/apache-conf.sh:
```diff
@@ -3,14 +3,17 @@
+# # Generate the /etc/httpd/conf.d/httpd-acm.conf file
 sudo mv /etc/httpd/conf.d/ssl.conf /etc/httpd/conf.d/ssl.conf-bkp
-printf "<VirtualHost *:443>\nServerName DOMAIN_NAME_PLACEHOLDER\nSSLEngine on\nSSLProtocol -all +TLSv1.2\nSSLCertificateKeyFile ""\nSSLCertificateFile ""\n</VirtualHost>\n" | sudo tee /etc/httpd/conf.d/httpd-acm.conf
+printf "Listen 443 https\nSSLPassPhraseDialog exec:/usr/libexec/httpd-ssl-pass-dialog\nSSLCryptoDevice pkcs11\n<VirtualHost *:443>\nServerName DOMAIN_NAME_PLACEHOLDER\nSSLEngine on\nSSLProtocol -all +TLSv1.2\nSSLCertificateKeyFile /etc/pki/tls/private/localhost.key\nSSLCertificateFile /etc/pki/tls/certs/localhost.crt\n</VirtualHost>\n" | sudo tee /etc/httpd/conf.d/httpd-acm.conf
```
I think this can be a lot easier to follow if we use heredoc. Something like this:
```
sudo tee /etc/httpd/conf.de/httpd-acm.conf << EOL
Listen 443 https
SSLPassPhraseDialog exec:/usr/libexec/httpd-ssl-pass-dialog
SSLCryptoDevice pkcs11
<VirtualHost *:443>
ServerName DOMAIN_NAME_PLACEHOLDER
SSLEngine on\nSSLProtocol -all +TLSv1.2
SSLCertificateKeyFile /etc/pki/tls/private/localhost.key\nSSLCertificateFile /etc/pki/tls/certs/localhost.crt
</VirtualHost>
EOL
```

foersleo commented on streamline/src/assets/user-data-scripts/AL2023/apache-conf.sh:
```diff
@@ -3,14 +3,17 @@
 
-# Start the ACM for Nitro Enclaves service
+# Edit the OpenSSL configuration /etc/pki/tls/openssl.cnf
+sudo sed -i '/ssl_conf = ssl_module/a\engines = engine_section\n\n[engine_section]\npkcs11 = pkcs11_section\n\n[ pkcs11_section ]\nengine_id = pkcs11\ninit = 1' /etc/pki/tls/openssl.cnf
```
I'd assume the order of the sections in the openssl.cnf does not really matter (?). Could we split this into two to make it easier to review. Something along the following could work I assume:
```
sudo sed -i '/ssl_conf = ssl_module' /etc/pki/tls/openssl.cnf

sudo tee -a /etc/pki/tls/openssl.cnf << EOF
a\engines = engine_section

[engine_section]
pkcs11 = pkcs11_section

[ pkcs11_section ]
engine_id = pkcs11
init = 1
EOF
```

Also, not quite sure about the `a\engines`. I could not find anything about that format and would assume that is an accident?

Belhoussine replied:
I was not sure whether the order matters, which is why I opted to keep the order specified in the manual installation guide, but I will try the suggested approach to see if it works.

For the `a\engines`, the `a` command is just used to append the `engines...` line at the end of the line containing `ssl_conf = ...`

foersleo commented on streamline/src/assets/user-data-scripts/private-cert-conf.sh:
```diff
@@ -0,0 +1,3 @@
+
```
I guess this file should also have a shebang to be on the safe side.

foersleo commented on streamline/README.md:
```diff
@@ -1,215 +1,97 @@
 
 ![nitro_enclaves_acm_streamline](assets/images/nitro_enclaves_acm_streamline.svg)
 
 ### 1. Certificate Stack (Step 1)
```
We probably want some comment above that the `Step x` here and below refer to the steps from the manual documentation. Could be confusing without that.

foersleo commented on streamline/docs/cdk-usage.md:
````diff
@@ -0,0 +1,160 @@
+Deploy the stacks
+
+```bash
+cdk deploy --all --require-approval never
````
I do not think the default should have the example with the `--require-approval never` flag set as a general note.

While this guide is aimed at advanced users, it might be taken verbatim by people without as much experience. An advanced user will know if they want to suppress the need for approvals, but inexperienced users might make security-relevant changes without learning about it.

Same goes for all the instances of this below.

foersleo commented on cdk/setup-tool:
```diff
@@ -0,0 +1,252 @@
+
+MY_VERSION="1.0"
+MY_NAME="acmne-setup-tool"
+MY_DESCRIPTION="ACM for Nitro Enclaves Setup Tool"
```
This one is unused.

foersleo commented on cdk/setup-tool:
```diff
@@ -0,0 +1,252 @@
+# Deploy command implementation
+run_deploy_command() {
+    local setup_name=""
+    local args="$@"
```
Throughout the script there is some un-cleanliness around usage of arrays and assigning it to strings.

Have a look at https://github.com/koalaman/shellcheck

It is a great tool to tease out small issues and provides good explanations through their Wiki (https://www.shellcheck.net/wiki).

foersleo commented on cdk/setup-tool:
```diff
@@ -0,0 +1,252 @@
+    
+    CLI_COMMAND="npx ts-node --prefer-ts-exts src/bin/cli.ts ${args}"
+    
+    cdk deploy --all --require-approval never --app \""${CLI_COMMAND}"\"
```
`--require-approval never` hidden from the unassuming user. The command may change security posture of resources in their AWS account, so please do not blanket approve all those changes without the user knowing or having a choice.

foersleo commented on cdk/setup-tool:
```diff
@@ -0,0 +1,252 @@
+parse_destroy_args() {
+    local filtered_args=()
+    local setup_name=""
+    local destroy_all=false
+    local destroy_certificate=false
+    local destroy_role=false
+    local destroy_instance=false
```
nit: Is the `destroy_all` flag not redundant? We could just set the three specific flags and have less hassle to recover the flags in `run_destroy_command`. It should also make that function less complex having one flag less to handle.

foersleo commented on cdk/setup-tool:
```diff
@@ -0,0 +1,252 @@
+#!/bin/bash
+
+MY_VERSION="1.0"
+MY_NAME="acmne-setup-tool"
```
nit: personally I would not use the `MY_` prefix, but something like `TOOL_` or none at all

foersleo reviewed (commented):
I really like that you did not cram all of this into one single commit. On the other hand I noticed that some of the changes here and there are related to each other in a way that one can not really stand without the other. Are you planning on squashing some of these together? If so have you thought about doing fixup commits for easier finding the connections between different commits?

meerd commented on cdk/setup-tool:
```diff
@@ -0,0 +1,244 @@
+#!/bin/bash
+
+VERSION="1.0"
+NAME="acmne-setup-tool"
```
This needs updating. @Belhoussine 
%s/acme-setup-tool/setup-tool

meerd commented on cdk/README.md:
```diff
@@ -0,0 +1,136 @@
+cd aws-nitro-enclaves-acm/cdk
+npm install
+cdk bootstrap aws://<AWS_ACCOUNT>/<AWS_REGION>
+chmod +x ./setup-tool
```
Why do we need this step? GIT can keep file permissions.

Belhoussine replied:
Not needed indeed, missed this detail.

meerd commented on cdk/README.md:
```diff
@@ -0,0 +1,136 @@
+git clone <repository-url>
+cd aws-nitro-enclaves-acm/cdk
+npm install
+cdk bootstrap aws://<AWS_ACCOUNT>/<AWS_REGION>
```
I'd suggest using `AWS-ACCOUNT-NUMBER` here for clarity.

Belhoussine replied:
Should I maybe use `AWS-ACCOUNT-ID` to be consistent with the rest of the documentation (which uses the term ID instead of Number)?

meerd commented on cdk/README.md:
```diff
@@ -0,0 +1,136 @@
+  --setup-name my-setup \
+  --aws-region <region> \
+  --aws-account-id <account-id> \ 
+  --is-private \
```
nit: How about using `--with-private-cert` here?

Belhoussine replied:
I think`--with-private-cert` might be a little confusing in the case where a customer already has a (private) ACM certificate (and provide the `existingCertificateArn` to our `setup-tool`). They would still need to provide a flag to describe if the certificate **is private** or not (hence the `--is-private` naming).
However, I agree that `--is-private` may not be very verbose. Should I maybe change it to `--is-private-cert` for more clarity?

meerd commented on cdk/README.md:
```diff
@@ -0,0 +1,136 @@
+  --certificate-arn arn:aws:acm:<region>:<account-id>:certificate/xxxx-yyy-zz \
+  --key-pair-name <key-pair-name> \
+  --instance-type <instance-type> \
+  --ami-type  <AL2|AL2023> \
```
Is this always choosing the latest available AL2/AL2023 AMI?

Belhoussine replied:
Yes indeed [`AmazonLinuxImage`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2.AmazonLinuxImage.html) selects the latest version of the chosen [`AmazonLinuxGeneration`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2.AmazonLinuxGeneration.html).

meerd commented on cdk/README.md:
````diff
@@ -0,0 +1,136 @@
+
+**Deployment with an existing ACM Certificate (public)**
+```bash
+./setup-tool deploy \
````
I suggest moving this tool to the /tools path.

Belhoussine replied:
Done

meerd commented on cdk/README.md:
````diff
@@ -0,0 +1,136 @@
+
+**Deployment with an existing ACM Certificate (public)**
+```bash
+./setup-tool deploy \
````
Would this command work if executed from the project root directory?

```
 ./cdk/setup-tool <arguments>
```

Belhoussine replied:
I implemented support for running the script from anywhere, by providing the path to it (as long as the `setup-tool` is in the `tools/` directory and the `cli.ts` is under the `cdk/src/bin/` directory.

Belhoussine committed:
fixup! cdk: Add user-data script for additional configuration if the certificate is private

Belhoussine committed:
cdk: Move setup-tool to parents tools/ folder, implement support to run the tool from location, handle support for --require-approval CDK flag

Belhoussine committed:
fixup! cdk: Fix httpd bug on AL2023 - resolve missing (undocumented) httpd-acm.conf directives and edit openssl.cnf
